### DISCUSSION TITLE
error in cload while changing firmware

### DISCUSSION THREAD

**Discussion by adarsh2798** [bitcraze/discussions#1233] [Q&A]
I am changing the firmware of crazyflie so that it can handle acceleration setpoint packets. I have installed wsl and cloned the repository by following instructions in https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/building-and-flashing/build/#debianubuntu. I am getting an error when i try to flash the firmware using cload ( see bottom of the post).

The changes i made in firmware are:

In crazyflie-firmware/src/modules/src, in the crtp_commander_generic.c file, I have made some changes.
```
/* ---===== 1 - packetType_e enum =====--- */ 
enum packet_type {
  stopType          = 0,
  velocityWorldType = 1,
  zDistanceType     = 2,
  cppmEmuType       = 3,
  altHoldType       = 4,
  hoverType         = 5,
  fullStateType     = 6,
  positionType      = 7,
  accType           = 8,                                      //edited(extended)
};
```

As shown, I am making changes for the commander.py class in cflib, to handle acceleration setpoints, so I have added ``accType=8`` as a new packet.
Then I add decoder function:
```
// ################################ EDITED(EXTENDED) PART##################################
/* accDecoder
 * Set the Crazyflie acceleration in the world coordinate system
 */
struct accPacket_s {
  float ax;        // m/s^2 in the world frame of reference
  float ay;        // ...
  float az;        // ...

} __attribute__((packed));
static void accDecoder(setpoint_t *setpoint, uint8_t type, const void *data, size_t datalen)
{const struct accPacket_s *values = data;

  ASSERT(datalen == sizeof(struct accPacket_s));

  setpoint->mode.x = modeAbs; // Mode for X acceleration
  setpoint->acceleration.x = values->ax;

  setpoint->mode.y = modeAbs; // Mode for Y acceleration
  setpoint->acceleration.y = values->ay;

  setpoint->mode.z = modeAbs; // Mode for Z acceleration
  setpoint->acceleration.z = values->az;
}

//#############################################################################################
```

then finally in packetDecoder array i add `` [accType] = accDecoder``
```
 /* ---===== 3 - packetDecoders array =====--- */
const static packetDecoder_t packetDecoders[] = {
  [stopType]          = stopDecoder,
  [velocityWorldType] = velocityDecoder,
  [zDistanceType]     = zDistanceDecoder,
  [cppmEmuType]       = cppmEmuDecoder,
  [altHoldType]       = altHoldDecoder,
  [hoverType]         = hoverDecoder,
  [fullStateType]     = fullStateDecoder,
  [positionType]      = positionDecoder,
  [accType]           = accDecoder,//edited(extended)
};
```

These are the only changes i made. Next I go to root directort of firmware and use
```
$ make cf2_defconfig
$ make -j 12
```

Then I press power button of crazyflie for 3 secodns for it to enter bootloader mode and I use the command:
```
make cload
```

And i get the following error:
```
adarsh@DESKTOP-N15S45F:~/crazyflie/firmware_v1/crazyflie-firmware$ make cload
tools/kbuild/Makefile.kbuild:147: warning: overriding recipe for target 'cload'
Makefile:136: warning: ignoring old recipe for target 'cload'
make[1]: Entering directory '/home/adarsh/crazyflie/firmware_v1/crazyflie-firmware/build'
python.exe -m cfloader  flash  cf2.bin stm32-fw
Traceback (most recent call last):
  File "C:\Python38\lib\runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Python38\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "c:\users\vadar\crazyflie-clients-python\src\cfloader\__main__.py", line 4, in <module>
    main()
  File "c:\users\vadar\crazyflie-clients-python\src\cfloader\__init__.py", line 117, in main
    targets.append(Target("cf2", target, type, [], []))
TypeError: __new__() takes 4 positional arguments but 6 were given
make[1]: *** [/home/adarsh/crazyflie/firmware_v1/crazyflie-firmware/Makefile:136: cload] Error 1
make[1]: Leaving directory '/home/adarsh/crazyflie/firmware_v1/crazyflie-firmware/build'
make: *** [tools/kbuild/Makefile.kbuild:150: sub-make] Error 2
```

I am using wsl on powershell.
I am new to all this. Any help would be appreciated.
Thanks.

**Comment by adarsh2798** [comment]
I switched to VM there i did not face any issues. Dont know about WSL though.

**Comment by ataffanel** [comment]
For WSL the problem you are having seems to come from a missmatched version between the client and the python lib. You can try to update both with "pip install --upgrade cfclient cflib". The new client (that implements the cload command line tool) does require an updated lib and the error you are getting looks very much what you would see in case of missmatch.

Moreover, WSL will need some special work to get the USB to work at all: by default Windows does not pass USB in WSL and a tool called USB-IP is required to make usb devices available in WSL. So this is something you need to make sure you have installed and setup correctly.
